File: internal/analyzers/couples/report_section.go
package couples

import (
	"fmt"
	"sort"
	"strconv"

	"github.com/Sumatoshi-tech/codefang/internal/analyzers/analyze"
)

// Section rendering constants.
const (
	ReportSectionTitle = "COUPLES"

	MetricTotalFiles      = "Total Files"
	MetricTotalDevelopers = "Total Developers"
	MetricTotalCoChanges  = "Total Co-Changes"
	MetricHighlyCoupled   = "Highly Coupled Pairs"
	MetricAvgCoupling     = "Avg Coupling"

	// DistStrongMin is the minimum coupling strength for "Strong" distribution bucket.
	DistStrongMin   = 0.7
	DistModerateMin = 0.4
	DistWeakMin     = 0.1
	DistLabelStrong = "Strong (>70%)"
	DistLabelMod    = "Moderate (40-70%)"
	DistLabelWeak   = "Weak (10-40%)"
	DistLabelNone   = "Minimal (<10%)"

	// IssueSeverityHighMin is the minimum coupling strength for "high" severity issues.
	IssueSeverityHighMin = 0.7
	IssueSeverityMedMin  = 0.4

	DefaultStatusMsg = "No coupling data available"
)

// ReportSection implements analyze.ReportSection for couples analysis.
type ReportSection struct {
	analyze.BaseReportSection

	metrics *ComputedMetrics
}

// NewReportSection creates a ReportSection from a couples report.
func NewReportSection(report analyze.Report) *ReportSection {
	if report == nil {
		report = analyze.Report{}
	}

	metrics, err := ComputeAllMetrics(report)
	if err != nil {
		metrics = &ComputedMetrics{}
	}

	score, msg := computeScore(metrics)

	return &ReportSection{
		BaseReportSection: analyze.BaseReportSection{
			Title:      ReportSectionTitle,
			Message:    msg,
			ScoreValue: score,
		},
		metrics: metrics,
	}
}

func computeScore(m *ComputedMetrics) (score float64, msg string) {
	if m.Aggregate.TotalFiles == 0 {
		return analyze.ScoreInfoOnly, DefaultStatusMsg
	}

	// Score is 1.0 - avg_coupling (lower coupling = better score).
	avg := m.Aggregate.AvgCouplingStrength
	score = 1.0 - avg

	if score < 0 {
		score = 0
	}

	const (
		goodThreshold = 0.7
		fairThreshold = 0.4
	)

	switch {
	case score >= goodThreshold:
		msg = fmt.Sprintf("Good - low coupling across %d files", m.Aggregate.TotalFiles)
	case score >= fairThreshold:
		msg = fmt.Sprintf("Fair - moderate coupling (%d highly coupled pairs)", m.Aggregate.HighlyCoupledPairs)
	default:
		msg = fmt.Sprintf("Poor - high coupling (%d highly coupled pairs need attention)", m.Aggregate.HighlyCoupledPairs)
	}

	return score, msg
}

// KeyMetrics returns the key metrics for the couples section.
func (s *ReportSection) KeyMetrics() []analyze.Metric {
	agg := s.metrics.Aggregate

	return []analyze.Metric{
		{Label: MetricTotalFiles, Value: strconv.Itoa(agg.TotalFiles)},
		{Label: MetricTotalDevelopers, Value: strconv.Itoa(agg.TotalDevelopers)},
		{Label: MetricTotalCoChanges, Value: strconv.FormatInt(agg.TotalCoChanges, 10)},
		{Label: MetricHighlyCoupled, Value: strconv.Itoa(agg.HighlyCoupledPairs)},
		{Label: MetricAvgCoupling, Value: fmt.Sprintf("%.0f%%", agg.AvgCouplingStrength*pctMultiplier)},
	}
}

// Distribution returns coupling strength distribution categories.
func (s *ReportSection) Distribution() []analyze.DistributionItem {
	couples := s.metrics.FileCoupling
	if len(couples) == 0 {
		return nil
	}

	counts := categorizeStrength(couples)
	total := len(couples)

	return []analyze.DistributionItem{
		{Label: DistLabelStrong, Percent: pct(counts.strong, total), Count: counts.strong},
		{Label: DistLabelMod, Percent: pct(counts.moderate, total), Count: counts.moderate},
		{Label: DistLabelWeak, Percent: pct(counts.weak, total), Count: counts.weak},
		{Label: DistLabelNone, Percent: pct(counts.minimal, total), Count: counts.minimal},
	}
}

// TopIssues returns the top N most coupled file pairs.
func (s *ReportSection) TopIssues(n int) []analyze.Issue {
	issues := s.buildSortedIssues()
	if n >= len(issues) {
		return issues
	}

	return issues[:n]
}

// AllIssues returns all coupled file pairs sorted by strength descending.
func (s *ReportSection) AllIssues() []analyze.Issue {
	return s.buildSortedIssues()
}

func (s *ReportSection) buildSortedIssues() []analyze.Issue {
	couples := s.metrics.FileCoupling
	if len(couples) == 0 {
		return nil
	}

	issues := make([]analyze.Issue, 0, len(couples))

	for _, cp := range couples {
		issues = append(issues, analyze.Issue{
			Name:     cp.File1 + " \u2194 " + cp.File2,
			Value:    fmt.Sprintf("%.0f%% (%d\u00d7)", cp.Strength*pctMultiplier, cp.CoChanges),
			Severity: severityForStrength(cp.Strength),
		})
	}

	// Sort by strength descending (highest coupling = most concerning = first).
	sort.Slice(issues, func(i, j int) bool {
		return issues[i].Value > issues[j].Value
	})

	return issues
}

// CreateReportSection creates a ReportSection from report data.
func (c *HistoryAnalyzer) CreateReportSection(report analyze.Report) analyze.ReportSection {
	return NewReportSection(report)
}

// --- Distribution helpers ---.

type strengthDistCounts struct {
	strong   int
	moderate int
	weak     int
	minimal  int
}

func categorizeStrength(couples []FileCouplingData) strengthDistCounts {
	var counts strengthDistCounts

	for _, cp := range couples {
		switch {
		case cp.Strength >= DistStrongMin:
			counts.strong++
		case cp.Strength >= DistModerateMin:
			counts.moderate++
		case cp.Strength >= DistWeakMin:
			counts.weak++
		default:
			counts.minimal++
		}
	}

	return counts
}

func severityForStrength(strength float64) string {
	switch {
	case strength >= IssueSeverityHighMin:
		return analyze.SeverityPoor
	case strength >= IssueSeverityMedMin:
		return analyze.SeverityFair
	default:
		return analyze.SeverityGood
	}
}

func pct(count, total int) float64 {
	if total == 0 {
		return 0
	}

	return float64(count) / float64(total)
}

File: crates/cf-couples/src/report_section.rs
//! Human-facing report-section logic.
//!
//! Computes the section score, status message, coupling-strength distribution,
//! and per-pair severity. These feed the (cosmetic, non-binding) terminal/HTML
//! summary; the numeric values follow the reference implementation so the
//! section behaves identically.

use crate::metrics::{ComputedMetrics, FileCouplingData};

/// Section title.
pub const REPORT_SECTION_TITLE: &str = "COUPLES";

/// Status message when no coupling data is available.
pub const DEFAULT_STATUS_MSG: &str = "No coupling data available";

/// Sentinel score used for info-only sections.
///
/// The canonical constant belongs in `cf-analyze`; the contractual value is
/// `-1.0`. Reconcile when `cf-analyze` lands (see crate TODOs).
pub const SCORE_INFO_ONLY: f64 = -1.0;

// Distribution thresholds.
const DIST_STRONG_MIN: f64 = 0.7;
const DIST_MODERATE_MIN: f64 = 0.4;
const DIST_WEAK_MIN: f64 = 0.1;

/// Distribution bucket labels.
pub const DIST_LABEL_STRONG: &str = "Strong (>70%)";
pub const DIST_LABEL_MOD: &str = "Moderate (40-70%)";
pub const DIST_LABEL_WEAK: &str = "Weak (10-40%)";
pub const DIST_LABEL_NONE: &str = "Minimal (<10%)";

// Issue severity thresholds.
const ISSUE_SEVERITY_HIGH_MIN: f64 = 0.7;
const ISSUE_SEVERITY_MED_MIN: f64 = 0.4;

/// Severity labels. The canonical string values belong in `cf-analyze`; the
/// contractual literals are reproduced here until that crate lands.
pub const SEVERITY_POOR: &str = "poor";
pub const SEVERITY_FAIR: &str = "fair";
pub const SEVERITY_GOOD: &str = "good";

/// Computes the section score and status message.
///
/// Score is `1.0 - avg_coupling_strength` (clamped to `[0, ∞)`); a higher score
/// means lower coupling. Returns [`SCORE_INFO_ONLY`] / [`DEFAULT_STATUS_MSG`]
/// when there are no files.
#[must_use]
pub fn compute_score(m: &ComputedMetrics) -> (f64, String) {
    const GOOD_THRESHOLD: f64 = 0.7;
    const FAIR_THRESHOLD: f64 = 0.4;
    if m.aggregate.total_files == 0 {
        return (SCORE_INFO_ONLY, DEFAULT_STATUS_MSG.to_string());
    }
    let mut score = 1.0 - m.aggregate.avg_coupling_strength;
    if score < 0.0 {
        score = 0.0;
    }
    let msg = if score >= GOOD_THRESHOLD {
        format!(
            "Good - low coupling across {} files",
            m.aggregate.total_files
        )
    } else if score >= FAIR_THRESHOLD {
        format!(
            "Fair - moderate coupling ({} highly coupled pairs)",
            m.aggregate.highly_coupled_pairs
        )
    } else {
        format!(
            "Poor - high coupling ({} highly coupled pairs need attention)",
            m.aggregate.highly_coupled_pairs
        )
    };
    (score, msg)
}

/// Coupling-strength distribution counts.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct StrengthDistCounts {
    pub strong: i32,
    pub moderate: i32,
    pub weak: i32,
    pub minimal: i32,
}

/// Categorizes coupling pairs by strength.
#[must_use]
pub fn categorize_strength(couples: &[FileCouplingData]) -> StrengthDistCounts {
    let mut counts = StrengthDistCounts::default();
    for cp in couples {
        if cp.strength >= DIST_STRONG_MIN {
            counts.strong += 1;
        } else if cp.strength >= DIST_MODERATE_MIN {
            counts.moderate += 1;
        } else if cp.strength >= DIST_WEAK_MIN {
            counts.weak += 1;
        } else {
            counts.minimal += 1;
        }
    }
    counts
}

/// Maps a coupling strength to a severity label.
#[must_use]
pub fn severity_for_strength(strength: f64) -> &'static str {
    if strength >= ISSUE_SEVERITY_HIGH_MIN {
        SEVERITY_POOR
    } else if strength >= ISSUE_SEVERITY_MED_MIN {
        SEVERITY_FAIR
    } else {
        SEVERITY_GOOD
    }
}

/// Computes a fraction `count / total`, returning `0.0` for an empty total.
#[must_use]
pub fn pct(count: i32, total: i32) -> f64 {
    if total == 0 {
        0.0
    } else {
        f64::from(count) / f64::from(total)
    }
}

#[cfg(test)]
#[allow(clippy::float_cmp)] // exact contract values (score constants, guards) are the point
mod tests {
    use super::*;
    use crate::metrics::AggregateData;

    fn metrics(total_files: i32, avg: f64, highly: i32) -> ComputedMetrics {
        ComputedMetrics {
            aggregate: AggregateData {
                total_files,
                avg_coupling_strength: avg,
                highly_coupled_pairs: highly,
                ..Default::default()
            },
            ..Default::default()
        }
    }

    #[test]
    fn score_no_files_is_info_only() {
        let (s, msg) = compute_score(&metrics(0, 0.0, 0));
        assert_eq!(s, SCORE_INFO_ONLY);
        assert_eq!(msg, DEFAULT_STATUS_MSG);
    }

    #[test]
    fn score_good_fair_poor() {
        let (s, msg) = compute_score(&metrics(5, 0.1, 0));
        assert!((s - 0.9).abs() < 1e-12);
        assert!(msg.starts_with("Good"));

        let (_s, msg) = compute_score(&metrics(5, 0.5, 2));
        assert!(msg.starts_with("Fair"));

        let (_s, msg) = compute_score(&metrics(5, 0.9, 3));
        assert!(msg.starts_with("Poor"));
    }

    #[test]
    fn distribution_categorization() {
        let couples = vec![
            FileCouplingData {
                file1: "a".into(),
                file2: "b".into(),
                co_changes: 1,
                strength: 0.8,
            },
            FileCouplingData {
                file1: "c".into(),
                file2: "d".into(),
                co_changes: 1,
                strength: 0.5,
            },
            FileCouplingData {
                file1: "e".into(),
                file2: "f".into(),
                co_changes: 1,
                strength: 0.2,
            },
            FileCouplingData {
                file1: "g".into(),
                file2: "h".into(),
                co_changes: 1,
                strength: 0.05,
            },
        ];
        let c = categorize_strength(&couples);
        assert_eq!(
            c,
            StrengthDistCounts {
                strong: 1,
                moderate: 1,
                weak: 1,
                minimal: 1
            }
        );
    }

    #[test]
    fn severity_mapping() {
        assert_eq!(severity_for_strength(0.8), SEVERITY_POOR);
        assert_eq!(severity_for_strength(0.5), SEVERITY_FAIR);
        assert_eq!(severity_for_strength(0.1), SEVERITY_GOOD);
    }

    #[test]
    fn pct_guards_zero() {
        assert_eq!(pct(0, 0), 0.0);
        assert_eq!(pct(1, 4), 0.25);
    }
}
